src/OrchardCore.Modules/OrchardCore.Deployment/Deployment/DeploymentPlanDeploymentSource.cs

```diff
@@ -1,16 +1,23 @@
+using System;
+using System.Collections.Generic;
 using System.Linq;
 using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
 using Newtonsoft.Json.Linq;
 
 namespace OrchardCore.Deployment.Deployment
 {
     public class DeploymentPlanDeploymentSource : IDeploymentSource
     {
         private readonly IDeploymentPlanService _deploymentPlanService;
+        private readonly IServiceProvider _serviceProvider;
 
-        public DeploymentPlanDeploymentSource(IDeploymentPlanService deploymentPlanService)
+        public DeploymentPlanDeploymentSource(
```

Comment:
Is there prevent from injecting `IEnumerable<IDeploymentStepFactory>>`?

Reply:
again they are transient so yes they are preventing from being injected by their nature.

Reply:
IDeploymentStepFactory is a singleton.

Reply:
And services.AddTransient<IDeploymentSource, DeploymentPlanDeploymentSource>();
So it will always have a different value.

Reply:
The main rule is that a singleton can't inject a scoped. A transient can inject anything but it is not recommanded that it injects a scoped, because the transient may be injected by a singleton (but it is then detected at runtime).

So, normally a transient `IDeploymentSource` can inject `IDeploymentStepFactory` singletons. What is wrong currently is that a transient `IDeploymentSource` injects a scoped `IDeploymentPlanService`, but that's okay as long as this transient is not injected by a sigleton.

Reply:
So I agree @jtkech 

I've reversed the resolutions, so that `IDeploymentPlanService` is resolved. Obviously not a major, as the transient isn't grabbed by a singleton, but it just felt better to fix.

Reply:
Okay cool

Reply:
@deanmarcussen 

When i said it's "wrong", i only meant that we have to take care of it

**So, if we think that this transient can't be injected by a singleton** (directly or indirectly), only by a scoped.

- This transient could still inject the scoped service, as we do with our tag helpers that can be considered as transients, see my comment https://github.com/OrchardCMS/OrchardCore/pull/8421#discussion_r566553905

    But it may be still useful to resolve the scoped service lazily by using the injected `IServiceProvider`, but here only if we want to resolve it on demand, only if we need it, but not to fix a DI lifetime issue.

    Note: As you know, injecting a service provider can also be used to solve a circular dependency

**If we think that this transient can be injected by a singleton** (directly or indirectly)

- We should resolve the scoped service lazily, but by using the right service provider, not by using the injected `IServiceProvider` that would be the one used when resolving the parent singleton.

    So, to fix a potential DI issue, we would have to retrieve the right service provider by using e.g. `RequestServices` which in fact, in our case, returns the service provider of the current shell scope.

So as you want, if we are in the first case we could just inject all services, or maybe just keep the code as you already did, maybe just remove or update the related comment. But in that case, why not registering these `IDeploymentSource` transients as scoped services instead. But if we are in the second case, using the injected service provider doesn't solve the DI concern. At least for now it seems that we are in the first case, otherwise we would already had DI issues

Reply:
Yeah I think we're probably on option 1 here, I'll have a look over the weekend. But probably revert to how it was.

`IDeploymentSoure` as scoped. I suspect this might make more sense. Will look further, but perhaps a different pr.

Reply:
Yes, no problem, anyway if a singleton reference e.g. a scoped through a transient, it is detected at runtime.

Reply:
Back to as before. 
`IDeploymentSource` could go to scoped, but it is not hurting anything currently. Bigger things to worry about ;)